File: EventMeenaAPI/src/EventMeena.Application/Services/Implementations/DocumentSigningService.cs
using AutoMapper;
using EventMeena.Application.DTOs.Common;
using EventMeena.Application.DTOs.DocumentSigning;
using EventMeena.Application.DTOs.Events;
using EventMeena.Application.Interfaces;
using EventMeena.Application.Services.Interfaces;
using EventMeena.Domain.Entities;
using EventMeena.Domain.Enums;
using System.Text.Json;

namespace EventMeena.Application.Services.Implementations;

/// <summary>
/// Document Signing Service Implementation
/// </summary>
public class DocumentSigningService : IDocumentSigningService
{
    private readonly IUnitOfWork _unitOfWork;
    private readonly IMapper _mapper;

    public DocumentSigningService(IUnitOfWork unitOfWork, IMapper mapper)
    {
        _unitOfWork = unitOfWork;
        _mapper = mapper;
    }

    #region Event Operations

    public async Task<ApiResponse<EventWithSignatureFieldsDto>> CreateDocumentEventAsync(
        Guid userId,
        string documentUrl,
        string documentFileName,
        CreateDocumentEventRequest request)
    {
        try
        {
            await _unitOfWork.BeginTransactionAsync();

            // إنشاء الحدث
            var evt = new Event
            {
                Id = Guid.NewGuid(),
                UserId = userId,
                Title = request.Title,
                Description = request.Description,
                Type = EventType.DocumentSigning,
                Status = EventStatus.Draft,
                DocumentUrl = documentUrl,
                DocumentFileName = documentFileName,
                CoverImage = request.CoverImage,
                ThemeColor = request.ThemeColor,
                Language = request.Language ?? "ar",
                StartDate = request.StartDate,
                EndDate = request.EndDate,
                RequireLogin = request.RequireLogin,
                AllowAnonymous = request.AllowAnonymous,
                MaxResponses = request.MaxResponses,
                AllowMultipleResponses = request.AllowMultipleResponses,
                IsPrivate = request.IsPrivate,
                AllowedEmailsJson = request.AllowedEmails != null
                    ? JsonSerializer.Serialize(request.AllowedEmails)
                    : null,
                AllowDownloadAfterSigning = request.AllowDownloadAfterSigning,
                SendCopyToSigner = request.SendCopyToSigner,
                ThankYouMessage = request.ThankYouMessage,
                ShareCode = GenerateShareCode()
            };

            await _unitOfWork.Events.AddAsync(evt);

            // إضافة حقول التوقيع
            var order = 0;
            foreach (var fieldRequest in request.SignatureFields)
            {
                var field = new SignatureField
                {
                    Id = Guid.NewGuid(),
                    EventId = evt.Id,
                    Label = fieldRequest.Label,
                    PageNumber = fieldRequest.PageNumber,
                    PositionX = fieldRequest.PositionX,
                    PositionY = fieldRequest.PositionY,
                    Width = fieldRequest.Width,
                    Height = fieldRequest.Height,
                    IsRequired = fieldRequest.IsRequired,
                    Order = order++,
                    FieldType = fieldRequest.FieldType,
                    IncludeDate = fieldRequest.IncludeDate,
                    IncludeName = fieldRequest.IncludeName
                };
                await _unitOfWork.SignatureFields.AddAsync(field);
            }

            await _unitOfWork.SaveChangesAsync();
            await _unitOfWork.CommitTransactionAsync();

            return await GetDocumentEventAsync(evt.Id, userId);
        }
        catch (Exception ex)
        {
            await _unitOfWork.RollbackTransactionAsync();
            return ApiResponse<EventWithSignatureFieldsDto>.FailureResponse($"حدث خطأ: {ex.Message}");
        }
    }

    public async Task<ApiResponse<EventWithSignatureFieldsDto>> CreateDocumentEventWithUrlAsync(
        Guid userId,
        CreateDocumentEventWithUrlRequest request)
    {
        try
        {
            await _unitOfWork.BeginTransactionAsync();

            // إنشاء الحدث
            var evt = new Event
            {
                Id = Guid.NewGuid(),
                UserId = userId,
                Title = request.Title,
                Description = request.Description,
                Type = EventType.DocumentSigning,
                Status = EventStatus.Published, // Published مباشرة
                DocumentUrl = request.DocumentUrl,
                DocumentFileName = request.DocumentFileName,
                Language = "ar",
                // إعدادات الوصول
                RequireLogin = request.RequireLogin,
                AllowAnonymous = !request.RequireLogin, // عكس RequireLogin
                IsPrivate = request.IsPrivate,
                AllowedEmailsJson = request.AllowedEmails != null && request.AllowedEmails.Any()
                    ? System.Text.Json.JsonSerializer.Serialize(request.AllowedEmails)
                    : null,
                // إعدادات التوقيع
                AllowDownloadAfterSigning = request.AllowDownloadAfterSigning,
                SendCopyToSigner = request.SendCopyToSigner,
                SignatureDisplayMode = request.SignatureDisplayMode ?? "inside",
                SigningMode = request.SigningMode ?? "single",
                ShareCode = GenerateShareCode()
            };

            await _unitOfWork.Events.AddAsync(evt);

            // إضافة حقول التوقيع
            var order = 0;
            foreach (var fieldRequest in request.SignatureFields)
            {
                var field = new SignatureField
                {
                    Id = Guid.NewGuid(),
                    EventId = evt.Id,
                    Label = fieldRequest.Label,
                    PageNumber = fieldRequest.PageNumber,
                    PositionX = fieldRequest.PositionX,
                    PositionY = fieldRequest.PositionY,
                    Width = fieldRequest.Width,
                    Height = fieldRequest.Height,
                    IsRequired = fieldRequest.IsRequired,
                    Order = fieldRequest.Order > 0 ? fieldRequest.Order : order++,
                    FieldType = fieldRequest.FieldType,
                    IncludeDate = fieldRequest.IncludeDate,
                    IncludeName = fieldRequest.IncludeName,
                    AssignedEmail = fieldRequest.AssignedEmail?.ToLowerInvariant()
                };
                await _unitOfWork.SignatureFields.AddAsync(field);
            }

            await _unitOfWork.SaveChangesAsync();
            await _unitOfWork.CommitTransactionAsync();

            return await GetDocumentEventAsync(evt.Id, userId);
        }
        catch (Exception ex)
        {
            await _unitOfWork.RollbackTransactionAsync();
            return ApiResponse<EventWithSignatureFieldsDto>.FailureResponse($"حدث خطأ: {ex.Message}");
        }
    }

    public async Task<ApiResponse<EventWithSignatureFieldsDto>> UpdateDocumentEventAsync(
        Guid eventId,
        Guid userId,
        UpdateDocumentEventRequest request)
    {
        var evt = await _unitOfWork.Events.GetByIdAsync(eventId);
        if (evt == null || evt.UserId != userId)
            return ApiResponse<EventWithSignatureFieldsDto>.FailureResponse("الحدث غير موجود");

        if (evt.Type != EventType.DocumentSigning)
            return ApiResponse<EventWithSignatureFieldsDto>.FailureResponse("هذا الحدث ليس من نوع توقيع الوثائق");

        try
        {
            await _unitOfWork.BeginTransactionAsync();

            // تحديث الحدث
            evt.Title = request.Title;
            evt.Description = request.Description;
            evt.CoverImage = request.CoverImage;
            evt.ThemeColor = request.ThemeColor;
            evt.Language = request.Language;
            evt.StartDate = request.StartDate;
            evt.EndDate = request.EndDate;
            evt.RequireLogin = request.RequireLogin;
            evt.AllowAnonymous = request.AllowAnonymous;
            evt.MaxResponses = request.MaxResponses;
            evt.AllowMultipleResponses = request.AllowMultipleResponses;
            evt.IsPrivate = request.IsPrivate;
            evt.AllowedEmailsJson = request.AllowedEmails != null
                ? JsonSerializer.Serialize(request.AllowedEmails)
                : null;
            evt.AllowDownloadAfterSigning = request.AllowDownloadAfterSigning;
            evt.SendCopyToSigner = request.SendCopyToSigner;
            evt.ThankYouMessage = request.ThankYouMessage;
            if (!string.IsNullOrEmpty(request.SignatureDisplayMode))
            {
                evt.SignatureDisplayMode = request.SignatureDisplayMode;
            }

            _unitOfWork.Events.Update(evt);

            // حذف حقول التوقيع القديمة وإضافة الجديدة
            await _unitOfWork.SignatureFields.DeleteByEventIdAsync(eventId);
            await _unitOfWork.SaveChangesAsync();

            var order = 0;
            foreach (var fieldRequest in request.SignatureFields)
            {
                var field = new SignatureField
                {
                    Id = fieldRequest.Id != Guid.Empty ? fieldRequest.Id : Guid.NewGuid(),
                    EventId = evt.Id,
                    Label = fieldRequest.Label,
                    PageNumber = fieldRequest.PageNumber,
                    PositionX = fieldRequest.PositionX,
                    PositionY = fieldRequest.PositionY,
                    Width = fieldRequest.Width,
                    Height = fieldRequest.Height,
                    IsRequired = fieldRequest.IsRequired,
                    Order = order++,
                    FieldType = fieldRequest.FieldType,
                    IncludeDate = fieldRequest.IncludeDate,
                    IncludeName = fieldRequest.IncludeName,
                    AssignedEmail = fieldRequest.AssignedEmail?.ToLowerInvariant()
                };
                await _unitOfWork.SignatureFields.AddAsync(field);
            }

            await _unitOfWork.SaveChangesAsync();
            await _unitOfWork.CommitTransactionAsync();

            return await GetDocumentEventAsync(evt.Id, userId);
        }
        catch (Exception ex)
        {
            await _unitOfWork.RollbackTransactionAsync();
            return ApiResponse<EventWithSignatureFieldsDto>.FailureResponse($"حدث خطأ: {ex.Message}");
        }
    }

    public async Task<ApiResponse<EventWithSignatureFieldsDto>> GetDocumentEventAsync(Guid eventId, Guid userId)
    {
        var evt = await _unitOfWork.Events.GetByIdAsync(eventId);
        if (evt == null || evt.UserId != userId)
            return ApiResponse<EventWithSignatureFieldsDto>.FailureResponse("الحدث غير موجود");

        var signatureFields = await _unitOfWork.SignatureFields.GetByEventIdAsync(eventId);

        var dto = _mapper.Map<EventWithSignatureFieldsDto>(evt);
        dto.SignatureFields = _mapper.Map<List<SignatureFieldDto>>(signatureFields);

        return ApiResponse<EventWithSignatureFieldsDto>.SuccessResponse(dto);
    }

    public async Task<ApiResponse<EventWithSignatureFieldsDto>> GetDocumentEventByShareCodeAsync(string shareCode)
    {
        var evt = await _unitOfWork.Events.GetByShareCodeAsync(shareCode);
        if (evt == null)
            return ApiResponse<EventWithSignatureFieldsDto>.FailureResponse("الحدث غير موجود");

        if (evt.Status != EventStatus.Published)
            return ApiResponse<EventWithSignatureFieldsDto>.FailureResponse("الحدث غير متاح حالياً");

        if (evt.Type != EventType.DocumentSigning)
            return ApiResponse<EventWithSignatureFieldsDto>.FailureResponse("هذا الحدث ليس من نوع توقيع الوثائق");

        var signatureFields = await _unitOfWork.SignatureFields.GetByEventIdAsync(evt.Id);

        var dto = _mapper.Map<EventWithSignatureFieldsDto>(evt);
        dto.SignatureFields = _mapper.Map<List<SignatureFieldDto>>(signatureFields);

        return ApiResponse<EventWithSignatureFieldsDto>.SuccessResponse(dto);
    }

    #endregion

    #region Signature Field Operations

    public async Task<ApiResponse<SignatureFieldDto>> AddSignatureFieldAsync(
        Guid eventId,
        Guid userId,
        CreateSignatureFieldRequest request)
    {
        var evt = await _unitOfWork.Events.GetByIdAsync(eventId);
        if (evt == null || evt.UserId != userId)
            return ApiResponse<SignatureFieldDto>.FailureResponse("الحدث غير موجود");

        if (evt.Type != EventType.DocumentSigning)
            return ApiResponse<SignatureFieldDto>.FailureResponse("هذا الحدث ليس من نوع توقيع الوثائق");

        var existingFields = await _unitOfWork.SignatureFields.GetByEventIdAsync(eventId);
        var maxOrder = existingFields.Any() ? existingFields.Max(f => f.Order) : -1;

        var field = new SignatureField
        {
            Id = Guid.NewGuid(),
            EventId = eventId,
            Label = request.Label,
            PageNumber = request.PageNumber,
            PositionX = request.PositionX,
            PositionY = request.PositionY,
            Width = request.Width,
            Height = request.Height,
            IsRequired = request.IsRequired,
            Order = maxOrder + 1,
            FieldType = request.FieldType,
            IncludeDate = request.IncludeDate,
            IncludeName = request.IncludeName
        };

        await _unitOfWork.SignatureFields.AddAsync(field);
        await _unitOfWork.SaveChangesAsync();

        return ApiResponse<SignatureFieldDto>.SuccessResponse(_mapper.Map<SignatureFieldDto>(field));
    }

    public async Task<ApiResponse<SignatureFieldDto>> UpdateSignatureFieldAsync(
        Guid fieldId,
        Guid userId,
        UpdateSignatureFieldRequest request)
    {
        var field = await _unitOfWork.SignatureFields.GetByIdAsync(fieldId);
        if (field == null)
            return ApiResponse<SignatureFieldDto>.FailureResponse("حقل التوقيع غير موجود");

        var evt = await _unitOfWork.Events.GetByIdAsync(field.EventId);
        if (evt == null || evt.UserId != userId)
            return ApiResponse<SignatureFieldDto>.FailureResponse("غير مصرح لك بتعديل هذا الحقل");

        field.Label = request.Label;
        field.PageNumber = request.PageNumber;
        field.PositionX = request.PositionX;
        field.PositionY = request.PositionY;
        field.Width = request.Width;
        field.Height = request.Height;
        field.IsRequired = request.IsRequired;
        field.FieldType = request.FieldType;
        field.IncludeDate = request.IncludeDate;
        field.IncludeName = request.IncludeName;

        _unitOfWork.SignatureFields.Update(field);
        await _unitOfWork.SaveChangesAsync();

        return ApiResponse<SignatureFieldDto>.SuccessResponse(_mapper.Map<SignatureFieldDto>(field));
    }

    public async Task<ApiResponse> DeleteSignatureFieldAsync(Guid fieldId, Guid userId)
    {
        var field = await _unitOfWork.SignatureFields.GetByIdAsync(fieldId);
        if (field == null)
            return ApiResponse.FailureResponse("حقل التوقيع غير موجود");

        var evt = await _unitOfWork.Events.GetByIdAsync(field.EventId);
        if (evt == null || evt.UserId != userId)
            return ApiResponse.FailureResponse("غير مصرح لك بحذف هذا الحقل");

        _unitOfWork.SignatureFields.Delete(field);
        await _unitOfWork.SaveChangesAsync();

        return ApiResponse.SuccessResponse("تم حذف حقل التوقيع بنجاح");
    }

    public async Task<ApiResponse<List<SignatureFieldDto>>> GetSignatureFieldsAsync(Guid eventId)
    {
        var fields = await _unitOfWork.SignatureFields.GetByEventIdAsync(eventId);
        return ApiResponse<List<SignatureFieldDto>>.SuccessResponse(
            _mapper.Map<List<SignatureFieldDto>>(fields));
    }

    #endregion

    #region Signature Operations

    public async Task<ApiResponse<DocumentSignatureDto>> SubmitSignatureAsync(
        Guid responseId,
        SubmitSignatureRequest request,
        string? ipAddress,
        string? userAgent)
    {
        var response = await _unitOfWork.Responses.GetByIdAsync(responseId);
        if (response == null)
            return ApiResponse<DocumentSignatureDto>.FailureResponse("الرد غير موجود");

        var field = await _unitOfWork.SignatureFields.GetByIdAsync(request.SignatureFieldId);
        if (field == null)
            return ApiResponse<DocumentSignatureDto>.FailureResponse("حقل التوقيع غير موجود");

        // التحقق من عدم وجود توقيع سابق
        var exists = await _unitOfWork.DocumentSignatures.ExistsAsync(request.SignatureFieldId, responseId);
        if (exists)
            return ApiResponse<DocumentSignatureDto>.FailureResponse("تم التوقيع على هذا الحقل مسبقاً");

        var signature = new DocumentSignature
        {
            Id = Guid.NewGuid(),
            SignatureFieldId = request.SignatureFieldId,
            ResponseId = responseId,
            SignerName = request.SignerName,
            SignerEmail = request.SignerEmail,
            SignerPhone = request.SignerPhone,
            SignatureData = request.SignatureData,
            SignedAt = DateTime.UtcNow,
            IpAddress = ipAddress,
            UserAgent = userAgent
        };

        await _unitOfWork.DocumentSignatures.AddAsync(signature);
        await _unitOfWork.SaveChangesAsync();

        return ApiResponse<DocumentSignatureDto>.SuccessResponse(_mapper.Map<DocumentSignatureDto>(signature));
    }

    public async Task<ApiResponse<List<DocumentSignatureDto>>> SubmitAllSignaturesAsync(
        SubmitAllSignaturesRequest request,
        string? ipAddress,
        string? userAgent)
    {
        try
        {
            await _unitOfWork.BeginTransactionAsync();

            // التحقق من وجود الحدث
            var evt = await _unitOfWork.Events.GetByIdAsync(request.EventId);
            if (evt == null)
                return ApiResponse<List<DocumentSignatureDto>>.FailureResponse("الحدث غير موجود");

            // إنشاء Response جديد إذا لم يكن موجوداً
            var responseId = request.ResponseId;
            if (responseId == Guid.Empty)
            {
                responseId = Guid.NewGuid();
                var response = new Response
                {
                    Id = responseId,
                    EventId = request.EventId,
                    RespondentName = request.SignerName,
                    RespondentEmail = request.SignerEmail,
                    RespondentIp = ipAddress,
                    UserAgent = userAgent,
                    Status = ResponseStatus.Completed,
                    StartedAt = DateTime.UtcNow,
                    CompletedAt = DateTime.UtcNow
                };
                await _unitOfWork.Responses.AddAsync(response);
            }

            var signatures = new List<DocumentSignature>();

            foreach (var sig in request.Signatures)
            {
                var field = await _unitOfWork.SignatureFields.GetByIdAsync(sig.SignatureFieldId);
                if (field == null)
                    continue;

                var exists = await _unitOfWork.DocumentSignatures.ExistsAsync(sig.SignatureFieldId, responseId);
                if (exists)
                    continue;

                var signature = new DocumentSignature
                {
                    Id = Guid.NewGuid(),
                    SignatureFieldId = sig.SignatureFieldId,
                    ResponseId = responseId,
                    SignerName = request.SignerName,
                    SignerEmail = request.SignerEmail,
                    SignerPhone = request.SignerPhone,
                    SignatureData = sig.SignatureData,
                    SignedAt = DateTime.UtcNow,
                    IpAddress = ipAddress,
                    UserAgent = userAgent
                };

                await _unitOfWork.DocumentSignatures.AddAsync(signature);
                signatures.Add(signature);
            }

            await _unitOfWork.SaveChangesAsync();
            await _unitOfWork.CommitTransactionAsync();

            return ApiResponse<List<DocumentSignatureDto>>.SuccessResponse(
                _mapper.Map<List<DocumentSignatureDto>>(signatures));
        }
        catch (Exception ex)
        {
            await _unitOfWork.RollbackTransactionAsync();
            return ApiResponse<List<DocumentSignatureDto>>.FailureResponse($"حدث خطأ: {ex.Message}");
        }
    }

    public async Task<ApiResponse<List<DocumentSignatureDto>>> GetEventSignaturesAsync(Guid eventId, Guid userId)
    {
        var evt = await _unitOfWork.Events.GetByIdAsync(eventId);
        if (evt == null || evt.UserId != userId)
            return ApiResponse<List<DocumentSignatureDto>>.FailureResponse("الحدث غير موجود");

        var signatures = await _unitOfWork.DocumentSignatures.GetByEventIdWithFieldInfoAsync(eventId);
        return ApiResponse<List<DocumentSignatureDto>>.SuccessResponse(
            _mapper.Map<List<DocumentSignatureDto>>(signatures));
    }

    public async Task<ApiResponse<List<DocumentSignatureDto>>> GetResponseSignaturesAsync(Guid responseId)
    {
        var signatures = await _unitOfWork.DocumentSignatures.GetByResponseIdAsync(responseId);
        return ApiResponse<List<DocumentSignatureDto>>.SuccessResponse(
            _mapper.Map<List<DocumentSignatureDto>>(signatures));
    }

    #endregion

    #region PDF Operations

    public async Task<ApiResponse> UpdateDocumentUrlAsync(Guid eventId, Guid userId, string documentUrl, string fileName)
    {
        var evt = await _unitOfWork.Events.GetByIdAsync(eventId);
        if (evt == null || evt.UserId != userId)
            return ApiResponse.FailureResponse("الحدث غير موجود");

        if (evt.Type != EventType.DocumentSigning)
            return ApiResponse.FailureResponse("هذا الحدث ليس من نوع توقيع الوثائق");

        evt.DocumentUrl = documentUrl;
        evt.DocumentFileName = fileName;

        _unitOfWork.Events.Update(evt);
        await _unitOfWork.SaveChangesAsync();

        return ApiResponse.SuccessResponse("تم تحديث الوثيقة بنجاح");
    }

    #endregion

    #region Private Methods

    private static string GenerateShareCode()
    {
        const string chars = "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghjkmnpqrstuvwxyz23456789";
        var random = new Random();
        return new string(Enumerable.Repeat(chars, 8)
            .Select(s => s[random.Next(s.Length)]).ToArray());
    }

    #endregion
}

File: event-meena/lib/api/services/documentSigningService.ts
/**
 * Document Signing Service - خدمة توقيع الوثائق
 * الاتصال بـ Backend API لتوقيع الوثائق
 */

import { apiClient, ApiResponse } from "../client";
import {
  DocumentSigningEvent,
  SignatureField,
  DocumentSignature,
  CreateDocumentEventRequest,
  CreateDocumentEventWithUrlRequest,
  UpdateDocumentEventRequest,
  CreateSignatureFieldRequest,
  UpdateSignatureFieldRequest,
  SubmitSignatureRequest,
  SubmitAllSignaturesRequest,
} from "@/types/document-signing";

// ============================================================
// Backend DTOs
// ============================================================

interface BackendSignatureFieldDto {
  id: string;
  eventId: string;
  label: string;
  pageNumber: number;
  positionX: number;
  positionY: number;
  width: number;
  height: number;
  isRequired: boolean;
  order: number;
  fieldType: number; // 1=Signature, 2=Initials, 3=Date, 4=Text, 5=Checkbox
  includeDate: boolean;
  includeName: boolean;
  assignedEmail: string | null; // إيميل الموقّع المخصص (multi-signer)
  createdAt: string;
  updatedAt: string;
}

interface BackendDocumentSignatureDto {
  id: string;
  signatureFieldId: string;
  responseId: string;
  signerName: string;
  signerEmail: string;
  signerPhone: string | null;
  signatureData: string;
  signedAt: string;
  fieldLabel: string | null;
  pageNumber: number | null;
}

interface BackendEventWithSignatureFieldsDto {
  id: string;
  title: string;
  description: string | null;
  type: number;
  status: number;
  shareCode: string;
  documentUrl: string | null;
  documentFileName: string | null;
  allowDownloadAfterSigning: boolean;
  sendCopyToSigner: boolean;
  signatureDisplayMode: string;
  signingMode: string; // نوع التوقيع: single أو multi
  signatureFields: BackendSignatureFieldDto[];
  // إعدادات الوصول
  requireLogin: boolean;
  isPrivate: boolean;
  allowedEmails: string[] | null;
  createdAt: string;
  updatedAt: string;
}

// ============================================================
// Mappers
// ============================================================

const fieldTypeMap: Record<number, SignatureField["fieldType"]> = {
  1: "signature",
  2: "initials",
  3: "date",
  4: "text",
  5: "checkbox",
};

const mapSignatureField = (backend: BackendSignatureFieldDto): SignatureField => ({
  id: backend.id,
  eventId: backend.eventId,
  label: backend.label,
  fieldType: fieldTypeMap[backend.fieldType] || "signature",
  pageNumber: backend.pageNumber,
  positionX: backend.positionX,
  positionY: backend.positionY,
  width: backend.width,
  height: backend.height,
  isRequired: backend.isRequired,
  order: backend.order,
  includeDate: backend.includeDate,
  includeName: backend.includeName,
  assignedEmail: backend.assignedEmail || undefined,
  createdAt: backend.createdAt,
  updatedAt: backend.updatedAt,
});

const mapDocumentSignature = (backend: BackendDocumentSignatureDto): DocumentSignature => ({
  id: backend.id,
  signatureFieldId: backend.signatureFieldId,
  responseId: backend.responseId,
  signerName: backend.signerName,
  signerEmail: backend.signerEmail,
  signerPhone: backend.signerPhone || undefined,
  signatureData: backend.signatureData,
  signedAt: backend.signedAt,
  fieldLabel: backend.fieldLabel || undefined,
  pageNumber: backend.pageNumber || undefined,
});

const mapDocumentSigningEvent = (backend: BackendEventWithSignatureFieldsDto): DocumentSigningEvent => ({
  id: backend.id,
  title: backend.title,
  description: backend.description || "",
  type: "document_signing",
  status: backend.status === 1 ? "draft" : backend.status === 2 ? "active" : "archived",
  shareCode: backend.shareCode,
  shareLink: `/e/${backend.shareCode}`,
  documentUrl: backend.documentUrl || "",
  documentFileName: backend.documentFileName || "",
  allowDownloadAfterSigning: backend.allowDownloadAfterSigning,
  sendCopyToSigner: backend.sendCopyToSigner,
  signatureDisplayMode: (backend.signatureDisplayMode as "inside" | "outside") || "inside",
  signingMode: (backend.signingMode as "single" | "multi") || "single",
  signatureFields: backend.signatureFields.map(mapSignatureField),
  sections: [],
  settings: {
    requireAuth: backend.requireLogin,
    isPrivate: backend.isPrivate,
    allowedEmails: backend.allowedEmails || undefined,
    allowAnonymous: !backend.requireLogin,
    allowMultipleResponses: false,
    showProgressBar: false,
    shuffleQuestions: false,
  },
  stats: {
    totalResponses: 0,
    completedResponses: 0,
    inProgressResponses: 0,
    completionRate: 0,
    averageTime: 0,
  },
  createdAt: backend.createdAt,
  updatedAt: backend.updatedAt,
});

// ============================================================
// API Functions
// ============================================================

/**
 * إنشاء حدث توقيع وثيقة جديد
 */
export async function createDocumentEvent(
  document: File,
  data: CreateDocumentEventRequest
): Promise<DocumentSigningEvent> {
  const formData = new FormData();
  formData.append("document", document);
  formData.append("title", data.title);
  if (data.description) formData.append("description", data.description);
  if (data.allowDownloadAfterSigning !== undefined) {
    formData.append("allowDownloadAfterSigning", String(data.allowDownloadAfterSigning));
  }
  if (data.sendCopyToSigner !== undefined) {
    formData.append("sendCopyToSigner", String(data.sendCopyToSigner));
  }

  const response = await apiClient.post<ApiResponse<BackendEventWithSignatureFieldsDto>>(
    "/DocumentSigning",
    formData,
    { headers: { "Content-Type": "multipart/form-data" } }
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "فشل إنشاء الحدث");
  }

  return mapDocumentSigningEvent(response.data.data);
}

/**
 * إنشاء حدث توقيع وثيقة جديد مع URL الوثيقة (بعد رفعها مسبقاً)
 */
export async function createDocumentEventWithUrl(
  data: CreateDocumentEventWithUrlRequest
): Promise<DocumentSigningEvent> {
  const response = await apiClient.post<ApiResponse<BackendEventWithSignatureFieldsDto>>(
    "/DocumentSigning/with-url",
    {
      title: data.title,
      description: data.description,
      documentUrl: data.documentUrl,
      documentFileName: data.documentFileName,
      signatureFields: data.signatureFields,
      // إعدادات الوصول (مفعّلة تلقائياً إذا لم تُحدد)
      requireLogin: data.requireLogin ?? true,
      isPrivate: data.isPrivate ?? true,
      allowedEmails: data.allowedEmails ?? [],
      // إعدادات التوقيع
      allowDownloadAfterSigning: data.allowDownloadAfterSigning,
      sendCopyToSigner: data.sendCopyToSigner,
      // طريقة عرض التوقيع للمشارك
      signatureDisplayMode: data.signatureDisplayMode ?? "inside",
      // نوع التوقيع
      signingMode: data.signingMode ?? "single",
    }
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "فشل إنشاء الحدث");
  }

  return mapDocumentSigningEvent(response.data.data);
}

/**
 * تحديث حدث توقيع
 */
export async function updateDocumentEvent(
  eventId: string,
  data: UpdateDocumentEventRequest
): Promise<DocumentSigningEvent> {
  const response = await apiClient.put<ApiResponse<BackendEventWithSignatureFieldsDto>>(
    `/DocumentSigning/${eventId}`,
    data
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "فشل تحديث الحدث");
  }

  return mapDocumentSigningEvent(response.data.data);
}

/**
 * الحصول على حدث توقيع
 */
export async function getDocumentEvent(eventId: string): Promise<DocumentSigningEvent> {
  const response = await apiClient.get<ApiResponse<BackendEventWithSignatureFieldsDto>>(
    `/DocumentSigning/${eventId}`
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "الحدث غير موجود");
  }

  return mapDocumentSigningEvent(response.data.data);
}

/**
 * الحصول على حدث توقيع برمز المشاركة (Public)
 */
export async function getDocumentEventByShareCode(shareCode: string): Promise<DocumentSigningEvent> {
  const response = await apiClient.get<ApiResponse<BackendEventWithSignatureFieldsDto>>(
    `/DocumentSigning/public/${shareCode}`
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "الحدث غير موجود");
  }

  return mapDocumentSigningEvent(response.data.data);
}

/**
 * تحديث ملف الوثيقة
 */
export async function updateDocument(eventId: string, document: File): Promise<void> {
  const formData = new FormData();
  formData.append("document", document);

  const response = await apiClient.put<ApiResponse<null>>(
    `/DocumentSigning/${eventId}/document`,
    formData,
    { headers: { "Content-Type": "multipart/form-data" } }
  );

  if (!response.data.success) {
    throw new Error(response.data.message || "فشل تحديث الوثيقة");
  }
}

// ============================================================
// Signature Fields
// ============================================================

/**
 * إضافة حقل توقيع
 */
export async function addSignatureField(
  eventId: string,
  data: CreateSignatureFieldRequest
): Promise<SignatureField> {
  const response = await apiClient.post<ApiResponse<BackendSignatureFieldDto>>(
    `/DocumentSigning/${eventId}/fields`,
    data
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "فشل إضافة حقل التوقيع");
  }

  return mapSignatureField(response.data.data);
}

/**
 * تحديث حقل توقيع
 */
export async function updateSignatureField(
  fieldId: string,
  data: UpdateSignatureFieldRequest
): Promise<SignatureField> {
  const response = await apiClient.put<ApiResponse<BackendSignatureFieldDto>>(
    `/DocumentSigning/fields/${fieldId}`,
    data
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "فشل تحديث حقل التوقيع");
  }

  return mapSignatureField(response.data.data);
}

/**
 * حذف حقل توقيع
 */
export async function deleteSignatureField(fieldId: string): Promise<void> {
  const response = await apiClient.delete<ApiResponse<null>>(
    `/DocumentSigning/fields/${fieldId}`
  );

  if (!response.data.success) {
    throw new Error(response.data.message || "فشل حذف حقل التوقيع");
  }
}

/**
 * الحصول على حقول التوقيع لحدث
 */
export async function getSignatureFields(eventId: string): Promise<SignatureField[]> {
  const response = await apiClient.get<ApiResponse<BackendSignatureFieldDto[]>>(
    `/DocumentSigning/${eventId}/fields`
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "فشل جلب حقول التوقيع");
  }

  return response.data.data.map(mapSignatureField);
}

// ============================================================
// Signatures
// ============================================================

/**
 * إرسال توقيع واحد
 */
export async function submitSignature(
  responseId: string,
  data: SubmitSignatureRequest
): Promise<DocumentSignature> {
  const response = await apiClient.post<ApiResponse<BackendDocumentSignatureDto>>(
    `/DocumentSigning/responses/${responseId}/sign`,
    data
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "فشل إرسال التوقيع");
  }

  return mapDocumentSignature(response.data.data);
}

/**
 * إرسال جميع التوقيعات مرة واحدة
 */
export async function submitAllSignatures(
  data: SubmitAllSignaturesRequest
): Promise<DocumentSignature[]> {
  // تحويل البيانات لتتوافق مع Backend
  // إذا لم يكن هناك responseId، نرسل Guid.Empty وسيقوم Backend بإنشاء Response جديد
  const backendData = {
    eventId: data.eventId,
    responseId: data.responseId || "00000000-0000-0000-0000-000000000000",
    signerName: data.signerName,
    signerEmail: data.signerEmail || "",
    signerPhone: data.signerPhone,
    signatures: data.signatures.map((sig) => ({
      signatureFieldId: sig.fieldId, // تحويل fieldId إلى signatureFieldId
      signatureData: sig.signatureData || "",
    })),
  };

  const response = await apiClient.post<ApiResponse<BackendDocumentSignatureDto[]>>(
    "/DocumentSigning/sign-all",
    backendData
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "فشل إرسال التوقيعات");
  }

  return response.data.data.map(mapDocumentSignature);
}

/**
 * الحصول على توقيعات حدث
 */
export async function getEventSignatures(eventId: string): Promise<DocumentSignature[]> {
  const response = await apiClient.get<ApiResponse<BackendDocumentSignatureDto[]>>(
    `/DocumentSigning/${eventId}/signatures`
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "فشل جلب التوقيعات");
  }

  return response.data.data.map(mapDocumentSignature);
}

/**
 * الحصول على توقيعات رد معين
 */
export async function getResponseSignatures(responseId: string): Promise<DocumentSignature[]> {
  const response = await apiClient.get<ApiResponse<BackendDocumentSignatureDto[]>>(
    `/DocumentSigning/responses/${responseId}/signatures`
  );

  if (!response.data.success || !response.data.data) {
    throw new Error(response.data.message || "فشل جلب التوقيعات");
  }

  return response.data.data.map(mapDocumentSignature);
}

// ============================================================
// Export Service Object
// ============================================================

export const documentSigningService = {
  // Event operations
  createDocumentEvent,
  createDocumentEventWithUrl,
  updateDocumentEvent,
  getDocumentEvent,
  getDocumentEventByShareCode,
  updateDocument,

  // Signature fields
  addSignatureField,
  updateSignatureField,
  deleteSignatureField,
  getSignatureFields,

  // Signatures
  submitSignature,
  submitAllSignatures,
  getEventSignatures,
  getResponseSignatures,
};

